cli/global_state.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use crate::compilers::CompiledModule;
+use crate::compilers::JsCompiler;
+use crate::compilers::JsonCompiler;
+use crate::compilers::TsCompiler;
+use crate::deno_dir;
+use crate::deno_error::permission_denied;
+use crate::file_fetcher::SourceFileFetcher;
+use crate::flags;
+use crate::metrics::Metrics;
+use crate::msg;
+use crate::ops::JsonOp;
+use crate::permissions::DenoPermissions;
+use crate::progress::Progress;
+use deno::ErrBox;
+use deno::Loader;
+use deno::ModuleSpecifier;
+use deno::PinnedBuf;
+use futures::Future;
+use serde_json::Value;
+use std;
+use std::env;
+use std::ops::Deref;
+use std::str;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+/// Holds state of the program and can be accessed by V8 isolate.
+pub struct ThreadSafeGlobalState(Arc<GlobalState>);
+
+#[cfg_attr(feature = "cargo-clippy", allow(stutter))]
+pub struct GlobalState {
+  pub main_module: Option<ModuleSpecifier>,
+  pub dir: deno_dir::DenoDir,
+  pub argv: Vec<String>,
+  pub flags: flags::DenoFlags,
+  pub permissions: DenoPermissions,
+  pub metrics: Metrics,
+  pub progress: Progress,
+  pub file_fetcher: SourceFileFetcher,
+  pub js_compiler: JsCompiler,
+  pub json_compiler: JsonCompiler,
+  pub ts_compiler: TsCompiler,
+}
+
+impl Clone for ThreadSafeGlobalState {
+  fn clone(&self) -> Self {
+    ThreadSafeGlobalState(self.0.clone())
+  }
+}
+
+impl Deref for ThreadSafeGlobalState {
+  type Target = Arc<GlobalState>;
+  fn deref(&self) -> &Self::Target {
+    &self.0
+  }
+}
+
+impl Loader for ThreadSafeGlobalState {
```

Comment:
I would have thought this is more of a per-worker thing...

Reply:
That is a good question. I think that `Resolver` should be per-worker (because only "main worker" should have import maps applied (https://github.com/WICG/import-maps/issues/157) and `DENO_DIR`/`file_fetcher` are defined once.

Reply:
I've changed my mind, it will be `Loader` trait that is implemented on the `ThreadSafeState` (so worker's state). And it will handle checking of permissions (so you can scope dynamic imports per worker :tada:) and eventually fetch module using file fetcher from global state :+1: